consensus/istanbul/proxy/proxy_engine.go

```diff
@@ -0,0 +1,175 @@
+			return err
+		}
+
+		if len(p.config.ProxyConfigs) > 0 {
+			p.ph.addProxies <- p.config.ProxyConfigs
+		}
+
+	}
+
+	p.logger.Info("Proxy engine started")
+	return nil
+}
+
+func (p *proxyEngine) Stop() error {
+	if p.backend.IsProxiedValidator() {
+		if err := p.ph.Stop(); err != nil {
+			return err
+		}
+	}
+
+	p.logger.Info("Proxy engine stopped")
+	return nil
+}
+
+func (p *proxyEngine) HandleMsg(peer consensus.Peer, msgCode uint64, payload []byte) (bool, error) {
+	if msgCode == istanbul.ValEnodesShareMsg {
+		// For now, handle this in a goroutine
+		go p.handleValEnodesShareMsg(peer, payload)
+		return true, nil
+	} else if msgCode == istanbul.FwdMsg {
+		return p.handleForwardMsg(peer, payload)
+	} else if msgCode == istanbul.ConsensusMsg {
+		return p.handleConsensusMsg(peer, payload)
+	} else if msgCode == istanbul.EnodeCertificateMsg {
+		go p.handleEnodeCertificateMsg(peer, payload)
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (p *proxyEngine) AddProxy(node, externalNode *enode.Node) error {
+	if p.backend.IsProxiedValidator() {
+		p.ph.addProxies <- []*istanbul.ProxyConfig{&istanbul.ProxyConfig{InternalNode: node, ExternalNode: externalNode}}
+	} else {
+		return ErrNodeNotProxiedValidator
+	}
+
+	return nil
+}
+
+func (p *proxyEngine) RemoveProxy(node *enode.Node) error {
+	if p.backend.IsProxiedValidator() {
+		p.ph.removeProxies <- []*enode.Node{node}
+	} else {
+		return ErrNodeNotProxiedValidator
+	}
+
+	return nil
+}
+
+func (p *proxyEngine) RegisterProxyPeer(proxyPeer consensus.Peer) error {
+	logger := p.logger.New("func", "RegisterProxyPeer")
+	if p.backend.IsProxiedValidator() {
+		if proxyPeer.PurposeIsSet(p2p.ProxyPurpose) {
+			logger.Info("Got new proxy peer", "proxyPeer", proxyPeer)
+			p.ph.addProxyPeer <- proxyPeer
+		} else {
+			logger.Error("Unauthorized connected peer to the proxied validator", "peerID", proxyPeer.Node().ID())
+			return errUnauthorizedProxiedValidator
+		}
+	}
+
+	return nil
+}
+
+func (p *proxyEngine) UnregisterProxyPeer(proxyPeer consensus.Peer) {
+	if p.backend.IsProxiedValidator() && proxyPeer.PurposeIsSet(p2p.ProxyPurpose) {
+		p.ph.removeProxyPeer <- proxyPeer
+	}
+}
+
+func (p *proxyEngine) RegisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer) {
+	// TODO: Does this need a lock?
+	if p.backend.IsProxy() {
+		p.proxiedValidator = proxiedValidatorPeer
+	}
+}
+
+func (p *proxyEngine) UnregisterProxiedValidatorPeer(proxiedValidatorPeer consensus.Peer) {
+	if p.backend.IsProxy() && p.proxiedValidator != nil && proxiedValidatorPeer.Node().ID() == p.proxiedValidator.Node().ID() {
+		p.proxiedValidator = nil
+	}
+}
+
+func (p *proxyEngine) GetValidatorProxyAssignments() (map[common.Address]*enode.Node, error) {
+	valProxyAssignments := make(map[common.Address]*enode.Node)
+
+	select {
+	case p.ph.proxyHandlerOpCh <- func(getValidatorAssignements func([]common.Address, []enode.ID) map[common.Address]*proxy) {
+		valAssignments := getValidatorAssignements(nil, nil)
```

Comment:
The select + proxyHandlerOp writer + proxyHandlerOpDone read should be extracted to it's own proxyHandler function